st2actions/st2actions/policies/concurrency.py

```diff
@@ -0,0 +1,94 @@
+from st2common.services import executions
+from st2common.util import action_db as action_utils
+
+
+LOG = logging.getLogger(__name__)
+
+
+class ConcurrencyPolicy(base.ResourcePolicy):
+
+    def __init__(self, policy_ref, policy_type, *args, **kwargs):
+        self.coordinator = coordination.get_coordinator()
+        self.threshold = 0
+        super(ConcurrencyPolicy, self).__init__(policy_ref, policy_type, *args, **kwargs)
+
+    def _apply(self, target):
+        # Get the count of scheduled instances of the action.
+        scheduled = action_access.LiveAction.count(
+            action=target.action, status=action_constants.LIVEACTION_STATUS_SCHEDULED)
+
+        # Get the count of running instances of the action.
+        running = action_access.LiveAction.count(
+            action=target.action, status=action_constants.LIVEACTION_STATUS_RUNNING)
+
+        count = scheduled + running
+
+        # Mark the execution as scheduled if threshold is not reached or delayed otherwise.
+        if count < self.threshold:
+            LOG.debug('There are %s instances of %s in scheduled or running status. '
+                      'Threshold of %s is not reached. Action execution will be scheduled.',
+                      count, target.action, self._policy_ref)
+            status = action_constants.LIVEACTION_STATUS_SCHEDULED
+        else:
+            LOG.debug('There are %s instances of %s in scheduled or running status. '
+                      'Threshold of %s is reached. Action execution will be delayed.',
+                      count, target.action, self._policy_ref)
+            status = action_constants.LIVEACTION_STATUS_DELAYED
+
+        # Update the status in the database but do not publish.
+        target = action_utils.update_liveaction_status(
+            status=status,
+            liveaction_id=target.id,
+            publish=False)
+
+        action_execution_db = executions.update_execution(target)
+
+        extra = {'action_execution_db': action_execution_db, 'liveaction_db': target}
+        LOG.audit('The status of action execution (id=%s) / live action (id=%s) '
+                  'is changed from %s to %s.', action_execution_db.id, target.id,
+                  target.status, status, extra=extra)
+
+        LOG.info('The status of action execution (id=%s) / live action (id=%s) '
+                 'is changed from %s to %s.', action_execution_db.id, target.id,
+                 target.status, status)
+
+        return target
+
+    def apply(self, target):
+        # Exit if target not in schedulable state.
+        if target.status != action_constants.LIVEACTION_STATUS_REQUESTED:
+            LOG.debug('The live action is not schedulable. %s', target)
+            return target
+
+        # Warn users that the coordination service is not configured.
+        if not coordination.configured():
+            LOG.warn('Coordination service is not configured. Policy enforcement is best effort.')
```

Comment:
Without a properly configured coordinator, in the case where there are multiple schedulers, we cannot guarantee the concurrency can be enforced.  The zake and file drivers for the coordinator works on the local server only.